feat: New route for values

CharlesNepote opened the pull request:
Get the unique values of a given property and the corresponding number of products.

This is a first answer to #221 but it does not manage different product types or categories (they don't currently exist in the data model).

CharlesNepote committed:
Add /values/{k} route

Get the unique values of a given property and the corresponding number of products

CharlesNepote committed:
Add tests

+ few doc + debug

CharlesNepote committed:
Update pull-request.yml

Fix issue related to permissions

CharlesNepote committed:
Update pull-request.yml

Fix perms

CharlesNepote committed:
Update pull-request.yml

CharlesNepote committed:
Allow external auth server

CharlesNepote committed:
Merge branch 'New-route-for-values' of https://github.com/cquest/folksonomy_api into New-route-for-values

CharlesNepote committed:
Fix PR workflow

create the directory before running the container

CharlesNepote committed:
Update pull-request.yml

Start postgres later

CharlesNepote committed:
Comments

CharlesNepote committed:
Merge branch 'New-route-for-values' of https://github.com/cquest/folksonomy_api into New-route-for-values

CharlesNepote commented:
Fix #223.

alexgarel commented on folksonomy/api.py:
```diff
@@ -554,6 +570,37 @@ async def keys_list(response: Response,
+    check_owner_user(user, owner, allow_anonymous=True)
+    k, _ = sanitize_data(k, None)
+    cur, timing = await db.db_exec("""
+        SELECT json_agg(j.j)::json FROM(
+            SELECT json_build_object(
+                'v', v,
+                'product_count', count(*)
+                ) as j
+            FROM folksonomy
+            WHERE owner=%s and k=%s
+            GROUP BY v) as j;
+        """,
```
The formating does not help reading.

Moreover I think we should:
1. order by product_count desc
2. limit the result to a reasonable amount (100 ?) this could be a parameter

```suggestion
        SELECT json_agg(j.j)::json
        FROM (
            SELECT json_build_object(
                'v', v,
                'product_count', count(*)
                ) as j
            FROM folksonomy
            WHERE owner=%s and k=%s
            GROUP BY v
            ORDER BY count(*)
            LIMIT 100
        ) as j;
        """,
```

CharlesNepote replied:
Ok. Should `LIMIT 100` be a default, or should it be a positive parameter (no parameter = not limit)?
I think this should not be a default, but only a parameter.

And I also think we should add a parameter for a filter.

What would fit best?

1. /values/{key}?limit=50&offset=50&q=kca
2. /values/{key}?page_size=100&page=2&q=kca
3. or simply /values/{key}?limit=50&q=kca

I guess I would prefer solution 3.

alexgarel replied:
1. I really really want 100 to be a default @CharlesNepote (we have enough perf problem to avoid getting new one, because people don't read the docs)
2. I really want to a have a max for this limit (I would suggest 1000) again to avoid perfs pittfall (people using 1000000 as parameter and putting our infrastructure down)
3. +1 for solution 3 for now

alexgarel commented on tests/test_main.py:
```diff
@@ -412,6 +412,17 @@ def test_keys_list(with_sample):
         ]
 
 
+def test_get_unique_values(with_sample):
+    with TestClient(app) as client:
+        response = client.get("/values/color")
+        assert response.status_code == 200
+        data = sorted(response.json(), key=lambda d: d["v"])
+        assert data == [
+            {'v': 'green', 'product_count': 1},
+            {'v': 'red', 'product_count': 2}
+        ]
```
Great that you added a test.

Could you eventually add a test with a non existing key ?

CharlesNepote replied:
Done + one test for each param

alexgarel reviewed (commented):
That's almost good to me, just added some propositions.

alexgarel reviewed (approved):
I approve, letting you do necessary changes.

CharlesNepote committed:
Better readability

CharlesNepote committed:
More exclusion

CharlesNepote committed:
/values limit and filter

Add &limit to limit the number of values returned (default: 50; max 1000)
Add &q to filter to values containing the query string

CharlesNepote committed:
Add tests for /values

* test non-existing property
* test param limit
* test param q

CharlesNepote committed:
Merge branch 'main' into New-route-for-values

CharlesNepote committed:
Fix tests

CharlesNepote committed:
Fix 2 tests

CharlesNepote committed:
Avoird hardcoded database name

CharlesNepote committed:
Script to launch tests

* drop existing test database
* create new test database
* run tests

CharlesNepote committed:
/values now can returns []

CharlesNepote committed:
Avoird securiity alert

CharlesNepote committed:
Merge branch 'main' into New-route-for-values
